src/services/ui/src/api/submissionService.ts

```diff
@@ -0,0 +1,147 @@
+  // TODO: Can this be a GET instead of POST w/ empty body
+  return await API.post("os", "/getUploadUrl", {
+    body: {},
+  });
+};
+
+/** Given an {@link AttachmentSchema} and array of pre signed URLs, this
+ * will return an array of objects that bundle a URL, s3 key, and s3 bucket
+ * with the file(s) to upload */
+const buildUploadRecipes = (
+  attachments: AttachmentsSchema,
+  validAttachmentKeys: string[],
+  preSignedUrls: PreSignedURL[]
+): UploadRecipe[] => {
+  return preSignedUrls.map((obj, idx) => ({
+    ...obj,
+    // Using validAttachmentKeys ensures we don't have any undefined value
+    data: attachments[validAttachmentKeys[idx]] as File[],
+    title: validAttachmentKeys[idx]
+  }));
+};
+
+/** Uploads one file if given a file and its meta from {@link getPreSignedUrl} */
+const uploadAttachment = async (file: File, preSignedURL: PreSignedURL) => {
+  return await fetch(preSignedURL.url, {
+    body: file,
+    method: "PUT",
+  });
+};
+
+/** Pass in an array of UploadRecipes and get a back-end compatible object
+ * to store attachment data */
+const buildAttachmentObject = (
+  recipes: UploadRecipe[]
+): OnemacAttachmentSchema[] => {
+  return recipes.map((r) => r.data.map(f => ({
+    key: r.key,
+    filename: f.name,
+    title: r.title,
+    bucket: r.bucket,
+    uploadDate: Date.now()
+  }) satisfies OnemacAttachmentSchema)).flat();
+};
+
+const buildSubmissionPayload = <T extends Record<string, unknown>>(
+  data: T,
+  variant: SubmissionVariant,
+  user: OneMacUser | undefined,
+  authority?: string,
+  attachments?: UploadRecipe[],
+) => {
+  switch (variant) {
+    case SubmissionVariant.ACTION:
+      return {
+        ...data,
+        attachments: attachments ? buildAttachmentObject(attachments) : null
+      };
+    case SubmissionVariant.INITIAL:
+      return {
+        ...data,
+        attachments: attachments ? buildAttachmentObject(attachments) : null,
+        // TODO: How much of this is common use in the foreseeable future
+        origin: "micro",
+        authority: authority,
+        state: (data.id as string).split("-")[0],
+        submitterEmail: user?.user?.email ?? "N/A",
+        submitterName:
+          `${user?.user?.given_name} ${user?.user?.family_name}` ?? "N/A"
+      };
+  }
+};
+
+const submitForm = async <T>(endpoint: SubmissionServiceEndpoint, data: T) =>
+  await API.post("os", endpoint, {
+    body: data,
+  });
+
+/** A useful interface for submitting form data to our submission service */
+export const submit = async <T extends Record<string, unknown>>({
+  data,
+  endpoint,
+  variant,
+  user,
+  authority
+}: SubmissionServiceParameters<T>): Promise<SubmissionServiceResponse> => {
+  if (data?.attachments) {
+    const attachments = data.attachments as AttachmentsSchema;
+    const validFilesetKeys = Object.keys(attachments).filter(key => attachments[key] !== undefined);
+    const preSignedURLs = await Promise.all(validFilesetKeys.map(() => getPreSignedUrl()));
```

Comment:
@mdial89f @13bfrancis We get one pre-signed URL object back per GROUP of files since every key can hold an array of 1 or more. This was copied functionality from the original approach, but this DOES mean every file in a group associated to a single key is uploaded with the same URL and given the same key. This seems like a flaw, is it?

I believe we should be getting a pre-signed URL per file, not per file group. That way every file has its own upload url and key. But this is making some assumptions about how the upload service and retrieval of a file works.